app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java

```diff
@@ -1,52 +1,54 @@
 package fr.free.nrw.commons.contributions;
 
-import android.text.TextUtils;
 import android.view.LayoutInflater;
 import android.view.ViewGroup;
 import androidx.annotation.NonNull;
-import androidx.recyclerview.widget.RecyclerView;
+import androidx.paging.PagedListAdapter;
+import androidx.recyclerview.widget.DiffUtil;
 import fr.free.nrw.commons.R;
 import fr.free.nrw.commons.media.MediaClient;
 import java.util.ArrayList;
 import java.util.List;
 
 /**
- * Represents The View Adapter for the List of Contributions  
+ * Represents The View Adapter for the List of Contributions
  */
-public class ContributionsListAdapter extends RecyclerView.Adapter<ContributionViewHolder> {
+public class ContributionsListAdapter extends
+    PagedListAdapter<Contribution, ContributionViewHolder> {
 
     private Callback callback;
     private final MediaClient mediaClient;
     private List<Contribution> contributions;
 
-    public ContributionsListAdapter(Callback callback,
-        MediaClient mediaClient) {
+    ContributionsListAdapter(final Callback callback,
+        final MediaClient mediaClient) {
+        super(DIFF_CALLBACK);
         this.callback = callback;
         this.mediaClient = mediaClient;
         contributions = new ArrayList<>();
+        setHasStableIds(true);
     }
 
-    /**
-     * Creates the new View Holder which will be used to display items(contributions)
-     * using the onBindViewHolder(viewHolder,position) 
-     */
-    @NonNull
-    @Override
-    public ContributionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
-        ContributionViewHolder viewHolder = new ContributionViewHolder(
-                LayoutInflater.from(parent.getContext())
-                        .inflate(R.layout.layout_contribution, parent, false), callback, mediaClient);
-        return viewHolder;
-    }
+
+    private static DiffUtil.ItemCallback<Contribution> DIFF_CALLBACK =
+        new DiffUtil.ItemCallback<Contribution>() {
+            // Concert details may have changed if reloaded from the database,
+            // but ID is fixed.
+            @Override
+            public boolean areItemsTheSame(Contribution oldConcert, Contribution newConcert) {
+                return oldConcert.getPageId() == newConcert.getPageId();
+            }
+
+            @Override
+            public boolean areContentsTheSame(Contribution oldConcert,
+                Contribution newConcert) {
+                return oldConcert.equals(newConcert);
```

Comment:
So you overrode equals to get the effect of `old.pageId==new.pageId` here by converting them to strings first. Couple problems
1. Quite surprising, could have effects at any other comparison site
1. Android studio will create hashCode, equals & toString methods for you, press `alt+insert`
1. no hashcode implementation was added

To make explicit your intention you could have said
```
            @Override
            public boolean areContentsTheSame(Contribution oldConcert,
                Contribution newConcert) {
                return areItemsTheSame(oldConcert, newConcert);
```

which is I think what the intention was.

This leads in to the larger problem of a contribution can't change its contents which they should be able to do I think? This could lead to some very confusing errors where the data has been updated and it just won't render

Reply:
Agreed. I didn't notice this while making the changes. I have updated the `hashcode` and `equals` implementation for `Media` and `Contribution` classes. It would be great if you could take a look.